### PR TITLE
[5.3] Add a simple transformer.

### PR DESCRIPTION
**Pull request by crynobone**
Would be useful to support simple transformer without requiring full-blown `thephpleague/fractal`. The idea is to provide basic transformer so we can have standard output when transforming single
model or a collection.

This is essentially useful for API response development since whenever we add new fields to say a Model, our API would then automatically has access to this information. It would be nice to have an explicit transformer to format the output.

Got this idea based on #14333

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Model.php [hunk 1]
```diff
@@ -2381,6 +2382,17 @@ public function setIncrementing($value)
...
 
     /**
+     * Transform each attribute in the model using a callback.
+     *
+     * @param  callable  $callback
+     * @return \Illuminate\Support\Fluent
+     */
+    public function transform(callable $callback)
+    {
+        return (new Fluent($this->toArray()))->transform($callback);
```
Comment (rkgrep) [inline src/Illuminate/Database/Eloquent/Model.php]: I don't like `toArray()` conversion. It usually causes lots of issues. Moreover, it ignores hidden attributes which is not always necessary.

Comment (crynobone) [inline src/Illuminate/Database/Eloquent/Model.php]: Would welcome any alternative suggestion

Comment (crynobone) [inline src/Illuminate/Database/Eloquent/Model.php]:
```php
return new Fluent($callback($this));
```

Comment (rkgrep) [inline src/Illuminate/Database/Eloquent/Model.php]: I wonder why is `Fluent` even needs to be used?
`league/fractal` package provides quite good set of tools for transforming objects to API output. Perhaps something like that would be much better.

Comment (crynobone) [inline src/Illuminate/Database/Eloquent/Model.php]: Which is why it's a simple transformer.

Fractal does more with formatting the input via transformer and then formatting the output with serializer. `Fluent` and `Collection` (which already support `transform()` method) offer simple json serializer out of the box.

Comment (rkgrep) [inline src/Illuminate/Database/Eloquent/Model.php]: Yes, I got the idea and it is good. Although I think that some `Transformer` contract would be useful.

**Comment by crynobone** [comment]
If @taylorotwell like the idea, we could come out with a Transformer abstract class.

```php
<?php

namespace Illuminate\Support;

abstract class Transformer 
{
    public function __invoke(...$parameters)
    {
        return $this->handle(...$parameters);
    }

    abstract public function handle(...$parameters);
}
```

I haven't test this but it `__invoke` should works fine with `callable` typehint, right?

```php
return App\User::find(1)->transform(new UserTransformer);
```

```php
return App\User::all()->transform(new UserTransformer);
```

If  #14333 merged.

```php
return App\User::paginate()->transform(new UserTransformer);
```

**Comment by taylorotwell** [comment]
I think this is probably something that should be discussed on master branch since 5.3 is so close to release.

**Comment by crynobone** [comment]
Sure 👍

**Pull request closed without merging**